File: dart/test/client_test.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:test/test.dart';
import 'package:wpush/wpush.dart';

void main() {
  http.Request? lastReq;

  Client clientWith(MockClientHandler handler) {
    lastReq = null;
    final mock = MockClient((request) async {
      lastReq = request;
      return handler(request);
    });
    return Client(apiKey: 'k', httpClient: mock);
  }

  test('send success joins channel and headers', () async {
    final c = clientWith((req) async {
      return http.Response(
        jsonEncode({'code': 0, 'message': 'success', 'data': '12345'}),
        200,
        headers: {'content-type': 'application/json'},
      );
    });
    final mid = await c.send(title: 't', content: 'c', channel: ['wechat', 'dingtalk']);
    expect(mid, '12345');
    expect(lastReq!.headers['user-agent'], kUserAgent);
    expect(lastReq!.headers['x-api-key'], 'k');
    expect(lastReq!.url.path, endsWith('/api/v1/send'));
    final body = jsonDecode(lastReq!.body) as Map<String, dynamic>;
    expect(body['apikey'], 'k');
    expect(body['channel'], 'wechat,dingtalk');
    expect(body['title'], 't');
  });

  test('send numeric data stringified', () async {
    final c = clientWith((req) async {
      return http.Response(jsonEncode({'code': 0, 'message': 'ok', 'data': 999}), 200);
    });
    expect(await c.send(title: 't'), '999');
  });

  test('reject bool code', () async {
    final c = clientWith((req) async {
      return http.Response(jsonEncode({'code': false, 'message': 'nope', 'data': null}), 200);
    });
    try {
      await c.send(title: 't');
      fail('expected');
    } on WPushException catch (e) {
      expect(e.code, isNull);
    }
  });

  test('api error 401', () async {
    final c = clientWith((req) async {
      return http.Response(jsonEncode({'code': 401, 'message': 'API Key error', 'data': null}), 200);
    });
    try {
      await c.send(title: 't');
      fail('expected');
    } on WPushException catch (e) {
      expect(e.code, 401);
    }
  });

  test('option topic conflict', () async {
    var called = false;
    final c = clientWith((req) async {
      called = true;
      return http.Response('{}', 200);
    });
    await expectLater(
      c.send(title: 't', option: 'ops', topicCode: 'topic1'),
      throwsA(isA<ValidationException>()),
    );
    expect(called, isFalse);
  });

  test('empty api key rejected', () {
    expect(() => Client(apiKey: ''), throwsA(isA<ValidationException>()));
  });

  test('query success', () async {
    final c = clientWith((req) async {
      return http.Response(
        jsonEncode({
          'code': 0,
          'message': 'success',
          'data': {'id': '123', 'status': 1, 'title': 't'},
        }),
        200,
      );
    });
    final out = await c.query('123');
    expect(out['id'], '123');
    expect(out['status'], 1);
    final body = jsonDecode(lastReq!.body) as Map<String, dynamic>;
    expect(body['id'], '123');
    expect(body['apikey'], 'k');
  });

  test('idempotency header', () async {
    final c = clientWith((req) async {
      return http.Response(jsonEncode({'code': 0, 'message': 'ok', 'data': '1'}), 200);
    });
    await c.send(title: 't', idempotencyKey: 'idem-1');
    expect(lastReq!.headers['x-idempotency-key'], 'idem-1');
  });

  test('sendMail success', () async {
    final c = clientWith((req) async {
      return http.Response(jsonEncode({'code': 0, 'message': 'success', 'data': 'mail-1'}), 200);
    });
    expect(await c.sendMail(to: 'a@b.com', title: 'hi', content: 'body'), 'mail-1');
    expect(lastReq!.url.path, endsWith('/api/v1/send_mail'));
    final body = jsonDecode(lastReq!.body) as Map<String, dynamic>;
    expect(body['to'], 'a@b.com');
  });

  test('sendCode success', () async {
    final c = clientWith((req) async {
      return http.Response(jsonEncode({'code': 0, 'message': 'success', 'data': 'code-1'}), 200);
    });
    expect(await c.sendCode(phone: '13800138000', code: '123456'), 'code-1');
    expect(lastReq!.url.path, endsWith('/api/v1/send_code'));
  });

  test('sendCode bad phone', () async {
    final c = clientWith((req) async => http.Response('{}', 200));
    await expectLater(
      c.sendCode(phone: '123', code: '123456'),
      throwsA(isA<ValidationException>()),
    );
  });

  test('queryRelay success', () async {
    final c = clientWith((req) async {
      return http.Response(
        jsonEncode({
          'code': 0,
          'message': 'success',
          'data': {'id': 'r1', 'status': 1},
        }),
        200,
      );
    });
    final out = await c.queryRelay('r1');
    expect(out['id'], 'r1');
    expect(lastReq!.url.path, endsWith('/api/v1/query_relay'));
  });

  test('isSuccessCode rejects bool', () {
    expect(isSuccessCode(false), isFalse);
    expect(isSuccessCode(true), isFalse);
    expect(isSuccessCode(0), isTrue);
    expect(isSuccessCode(0.0), isTrue);
  });
}

File: go/client_test.go
package wpush

import (
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestSendSuccess(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/send" {
			t.Fatalf("path %s", r.URL.Path)
		}
		if r.Header.Get("User-Agent") != userAgent {
			t.Fatalf("ua %s", r.Header.Get("User-Agent"))
		}
		if r.Header.Get("X-API-Key") != "k" {
			t.Fatalf("key header")
		}
		body, _ := io.ReadAll(r.Body)
		var m map[string]any
		_ = json.Unmarshal(body, &m)
		if m["apikey"] != "k" || m["channel"] != "wechat,dingtalk" {
			t.Fatalf("body %#v", m)
		}
		_ = json.NewEncoder(w).Encode(map[string]any{"code": 0, "message": "success", "data": "12345"})
	}))
	defer ts.Close()
	c, err := NewClient(Options{APIKey: "k", BaseURL: ts.URL})
	if err != nil {
		t.Fatal(err)
	}
	id, err := c.Send(SendParams{Title: "t", Content: "c", Channel: []string{"wechat", "dingtalk"}})
	if err != nil || id != "12345" {
		t.Fatalf("id=%s err=%v", id, err)
	}
}

func TestRejectBoolCode(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		_ = json.NewEncoder(w).Encode(map[string]any{"code": false, "message": "nope", "data": nil})
	}))
	defer ts.Close()
	c, _ := NewClient(Options{APIKey: "k", BaseURL: ts.URL})
	_, err := c.Send(SendParams{Title: "t"})
	if err == nil {
		t.Fatal("expected error")
	}
	if e, ok := err.(*Error); !ok || e.Code != nil {
		t.Fatalf("got %#v", err)
	}
}

func TestAPIError401(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		_ = json.NewEncoder(w).Encode(map[string]any{"code": 401, "message": "bad key", "data": nil})
	}))
	defer ts.Close()
	c, _ := NewClient(Options{APIKey: "bad", BaseURL: ts.URL})
	_, err := c.Send(SendParams{Title: "t"})
	e, ok := err.(*Error)
	if !ok || e.Code == nil || *e.Code != 401 {
		t.Fatalf("%#v", err)
	}
}

func TestOptionTopicConflict(t *testing.T) {
	c, _ := NewClient(Options{APIKey: "k", BaseURL: "http://example.invalid"})
	_, err := c.Send(SendParams{Title: "t", Option: "ops", TopicCode: "x"})
	if _, ok := err.(*ValidationError); !ok {
		t.Fatalf("%#v", err)
	}
}

func TestQuerySuccess(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/query" {
			t.Fatalf("path %s", r.URL.Path)
		}
		body, _ := io.ReadAll(r.Body)
		var m map[string]any
		_ = json.Unmarshal(body, &m)
		if m["id"] != "123" || m["apikey"] != "k" {
			t.Fatalf("%#v", m)
		}
		_ = json.NewEncoder(w).Encode(map[string]any{
			"code": 0, "message": "success",
			"data": map[string]any{"id": "123", "status": float64(1)},
		})
	}))
	defer ts.Close()
	c, _ := NewClient(Options{APIKey: "k", BaseURL: ts.URL})
	out, err := c.Query("123", "")
	if err != nil || out["id"] != "123" {
		t.Fatalf("%v %#v", err, out)
	}
}

func TestSendNumericData(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte(`{"code":0,"message":"ok","data":999}`))
	}))
	defer ts.Close()
	c, _ := NewClient(Options{APIKey: "k", BaseURL: ts.URL})
	id, err := c.Send(SendParams{Title: "t"})
	if err != nil || id != "999" {
		t.Fatalf("id=%s err=%v", id, err)
	}
}

func TestSendMailSuccess(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/send_mail" {
			t.Fatalf("path %s", r.URL.Path)
		}
		body, _ := io.ReadAll(r.Body)
		var m map[string]any
		_ = json.Unmarshal(body, &m)
		if m["to"] != "a@b.com" || m["title"] != "hi" || m["content"] != "body" || m["apikey"] != "k" {
			t.Fatalf("body %#v", m)
		}
		_ = json.NewEncoder(w).Encode(map[string]any{"code": 0, "message": "success", "data": "mail-1"})
	}))
	defer ts.Close()
	c, _ := NewClient(Options{APIKey: "k", BaseURL: ts.URL})
	id, err := c.SendMail(SendMailParams{To: "a@b.com", Title: "hi", Content: "body"})
	if err != nil || id != "mail-1" {
		t.Fatalf("id=%s err=%v", id, err)
	}
}

func TestSendCodeSuccess(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/send_code" {
			t.Fatalf("path %s", r.URL.Path)
		}
		body, _ := io.ReadAll(r.Body)
		var m map[string]any
		_ = json.Unmarshal(body, &m)
		if m["phone"] != "13800138000" || m["code"] != "123456" {
			t.Fatalf("body %#v", m)
		}
		_ = json.NewEncoder(w).Encode(map[string]any{"code": 0, "message": "success", "data": "code-1"})
	}))
	defer ts.Close()
	c, _ := NewClient(Options{APIKey: "k", BaseURL: ts.URL})
	id, err := c.SendCode(SendCodeParams{Phone: "13800138000", Code: "123456"})
	if err != nil || id != "code-1" {
		t.Fatalf("id=%s err=%v", id, err)
	}
}

func TestSendCodeBadPhone(t *testing.T) {
	c, _ := NewClient(Options{APIKey: "k", BaseURL: "http://example.invalid"})
	_, err := c.SendCode(SendCodeParams{Phone: "123", Code: "123456"})
	if _, ok := err.(*ValidationError); !ok {
		t.Fatalf("%#v", err)
	}
}

func TestQueryRelaySuccess(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/query_relay" {
			t.Fatalf("path %s", r.URL.Path)
		}
		body, _ := io.ReadAll(r.Body)
		var m map[string]any
		_ = json.Unmarshal(body, &m)
		if m["id"] != "r1" || m["apikey"] != "k" {
			t.Fatalf("%#v", m)
		}
		_ = json.NewEncoder(w).Encode(map[string]any{
			"code": 0, "message": "success",
			"data": map[string]any{"id": "r1", "status": float64(1)},
		})
	}))
	defer ts.Close()
	c, _ := NewClient(Options{APIKey: "k", BaseURL: ts.URL})
	out, err := c.QueryRelay("r1", "")
	if err != nil || out["id"] != "r1" {
		t.Fatalf("%v %#v", err, out)
	}
}
